hashes/src/internal_macros.rs

```diff
@@ -84,6 +84,20 @@ macro_rules! hash_trait_impls {
             pub fn backward_hex(&self) -> impl '_ + core::fmt::LowerHex + core::fmt::UpperHex {
                 internals::hex::display::DisplayArray::<_, [u8; $bits / 8 * 2]>::new(self.0.iter().rev())
             }
+
+            /// Zero cost conversion between a fixed length byte array shared reference and
+            /// a shared reference to this Hash type.
+            pub fn from_bytes_ref(bytes: &[u8; $bits / 8]) -> &Self {
+                // Safety: Sound because Self is #[repr(transparent)] containing [u8; $bits / 8]
+                unsafe { &*(bytes as *const _ as *const Self) }
+            }
+
+            /// Zero cost conversion between a fixed length byte array exclusive reference and
+            /// an exclusive reference to this Hash type.
+            pub fn from_bytes_mut(bytes: &mut [u8; $bits / 8]) -> &mut Self {
+                // Safety: Sound because Self is #[repr(transparent)] containing [u8; $bits / 8]
+                unsafe { &mut *(bytes as *mut _ as *mut Self) }
+            }
```

Comment:
I think this is technically unsound because someone could call the macro outside of `hash_newtype` and get this functions implemented. Maybe I'm paranoid but I'd rather have it in `hash_type` macro.
And while you're at it add a non-doc comment: `#[repr(transaparent)] // required for soundness of from_bytes_ref/mut`, so that nobody removes it.

Also it'd be nice to document what `#[repr(transparent)]` means for consumers: The memory layout is guaranteed to be the same as `[u8; LEN]`.

Reply:
Ooh, good catch! I dutifully checked that every caller of this macro had `#[repr(transparent)]` (and was also surprised, because this isn't documented and it's actually hard to find the exact definitions of everything when everything is called `Hash` and defined in a macro soup). But of course, we are not necessarily the only callers of the macro.